### PR TITLE
[2.0 API] add SyncBatchNorm.convert_sync_batch_norm

### PR DESCRIPTION
**Pull request by ceci3**
### PR types

Others

### PR changes

APIs

### Describe

add convert_sync_batch_norm for SyncBatchNorm to convert standard bn to syncbn
fix batch norm when weight_attr=False or bias_attr=False

### CODE REVIEW

File: python/paddle/nn/layer/norm.py [hunk 1]
```diff
@@ -1112,3 +1113,45 @@ def forward(self, x):
...
+    @classmethod
+    def convert_sync_batchnorm(cls, layer):
+        """
+        Helper function to convert :class: `paddle.nn.BatchNorm*d` layers in the model to :class: `paddle.nn.SyncBatchNorm` layers.
+
+        Parameters:
+            layer(paddle.fluid.dygraph.Layer): model containing one or more `BatchNorm*d` layers.
+
+        Returns:
+            A new SyncBatchNorm layer object if origin layer is BatchNorm*d Layer.
```
Comment (wanghaoshuang) [inline python/paddle/nn/layer/norm.py]: 这里的表述是不是有问题？返回的不一定是SyncBatchNorm layer吧？

**Comment by wanghaoshuang** [comment]
PR标题有点太简单了。

File: python/paddle/nn/layer/norm.py (hunk 1 above)
Comment (ceci3) [reply python/paddle/nn/layer/norm.py]: 已修改, 返回的应该是原始传进来的模型

File: python/paddle/nn/layer/norm.py [hunk 2]
```diff
@@ -1112,3 +1113,45 @@ def forward(self, x):
...
             type="sync_batch_norm", inputs=inputs, outputs=outputs, attrs=attrs)
         return sync_batch_norm_out
+
+    @classmethod
+    def convert_sync_batchnorm(cls, layer):
+        """
+        Helper function to convert :class: `paddle.nn.BatchNorm*d` layers in the model to :class: `paddle.nn.SyncBatchNorm` layers.
+
+        Parameters:
+            layer(paddle.fluid.dygraph.Layer): model containing one or more `BatchNorm*d` layers.
```
Comment (jzhang533) [inline python/paddle/nn/layer/norm.py]: use `paddle.nn.Layer` would be better.

File: python/paddle/nn/layer/norm.py [hunk 3]
```diff
@@ -1112,3 +1113,45 @@ def forward(self, x):
...
+        """
+        layer_output = layer
+        if isinstance(layer, _BatchNormBase):
+            layer_output = SyncBatchNorm(layer._num_features, layer._epsilon,
+                                         layer._momentum, layer._weight_attr,
+                                         layer._bias_attr, layer._data_format,
+                                         layer._name)
+
+            if layer._weight_attr != False and layer._bias_attr != False:
+                with no_grad():
```
Comment (willthefrog) [inline python/paddle/nn/layer/norm.py:1103]: is this required? also possible to `__new__` a SyncBatchNorm, so that the weight does not need be created then discarded?

File: python/paddle/nn/layer/norm.py (hunk 2 above)
Comment (ceci3) [reply python/paddle/nn/layer/norm.py]: done, thanks

File: python/paddle/nn/layer/norm.py (hunk 3 above)
Comment (ceci3) [reply python/paddle/nn/layer/norm.py:1103]: done, thanks

File: python/paddle/nn/layer/norm.py [hunk 4]
```diff
@@ -566,17 +566,30 @@ def __init__(self,
...
+        if weight_attr == False:
+            self.weight = self.create_parameter(
+                attr=None, shape=param_shape, default_initializer=Constant(1.0))
+            self.weight.stop_gradient = True
+        else:
+            self.weight = self.create_parameter(
+                attr=self._weight_attr,
+                shape=param_shape,
+                default_initializer=Constant(1.0))
+            self.weight.stop_gradient = (self._weight_attr is False) or (
```
Comment (frankwhzhang) [inline python/paddle/nn/layer/norm.py]: self._weight_attr is False是多余的吧

File: python/paddle/nn/layer/norm.py [hunk 5]
```diff
@@ -566,17 +566,30 @@ def __init__(self,
...
+            self.bias = self.create_parameter(
+                attr=None,
+                shape=param_shape,
+                default_initializer=Constant(0.0),
+                is_bias=True)
+            self.bias.stop_gradient = True
+        else:
+            self.bias = self.create_parameter(
+                attr=self._bias_attr, shape=param_shape, is_bias=True)
+            self.bias.stop_gradient = (self._bias_attr is False) or (
```
Comment (frankwhzhang) [inline python/paddle/nn/layer/norm.py]: 同上

File: python/paddle/nn/layer/norm.py (hunk 4 above)
Comment (ceci3) [reply python/paddle/nn/layer/norm.py]: 已删掉，感谢~

File: python/paddle/nn/layer/norm.py [hunk 6]
```diff
@@ -1112,3 +1132,49 @@ def forward(self, x):
...
+                model = nn.Sequential(nn.Conv2d(3, 5, 3), nn.BatchNorm2d(5))
+                sync_model = nn.SyncBatchNorm.convert_sync_batchnorm(model)
+
+        """
+        layer_output = layer
+        if isinstance(layer, _BatchNormBase):
+            layer_output = SyncBatchNorm.__new__(SyncBatchNorm)
+            super(SyncBatchNorm, layer_output).__init__()
+
+            with no_grad():
```
Comment (willthefrog) [inline python/paddle/nn/layer/norm.py]: why is `no_grad` needed?

**Review by willthefrog — CHANGES_REQUESTED** [review]

File: python/paddle/nn/layer/norm.py (hunk 6 above)
Comment (ceci3) [reply python/paddle/nn/layer/norm.py]: 没有no_grad的话，assign这个流程好像会有梯度回传之类的问题

Comment (willthefrog) [reply python/paddle/nn/layer/norm.py]: this will be a simple python variable, should not involve gradient, anyway, doesn't hurt.

**Review by willthefrog — DISMISSED** [review]
LGTM

**Review by willthefrog — APPROVED** [review]
LGTM

**Review by jzhang533 — APPROVED** [review]
lgtm

File: python/paddle/nn/layer/norm.py [hunk 7]
```diff
@@ -566,17 +567,28 @@ def __init__(self,
         param_shape = [num_features]
 
         # create parameter
-        self.weight = self.create_parameter(
-            attr=self._weight_attr,
-            shape=param_shape,
-            default_initializer=Constant(1.0))
-        self.weight.stop_gradient = (self._weight_attr is False) or (
-            self._weight_attr and self._weight_attr.learning_rate == 0.)
+        if weight_attr == False:
```
Comment (XiaoguangHu01) [inline python/paddle/nn/layer/norm.py:570]: False -> None
None != False
看起来这个if-else逻辑跟下面这种写法等价
self.weight = self.create_parameter(
            attr=weight_attr,
            shape=param_shape,
            default_initializer=Constant(1.0))
self.weight.stop_gradient = (weight_attr == None) or (weight_attr.learning_rate == 0.)

File: python/paddle/nn/layer/norm.py [hunk 8]
```diff
@@ -566,17 +567,28 @@ def __init__(self,
...
+                attr=self._weight_attr,
+                shape=param_shape,
+                default_initializer=Constant(1.0))
+            self.weight.stop_gradient = self._weight_attr != None and self._weight_attr.learning_rate == 0.
 
-        self.bias = self.create_parameter(
-            attr=self._bias_attr, shape=param_shape, is_bias=True)
-        self.bias.stop_gradient = (self._bias_attr is False) or (
-            self._bias_attr and self._bias_attr.learning_rate == 0.)
+        if bias_attr == False:
```
Comment (XiaoguangHu01) [inline python/paddle/nn/layer/norm.py:581]: False -> None

File: python/paddle/nn/layer/norm.py [hunk 9]
```diff
@@ -1112,3 +1069,45 @@ def forward(self, x):
...
+
+        """
+        layer_output = layer
+        if isinstance(layer, _BatchNormBase):
+            layer_output = SyncBatchNorm(layer._num_features, layer._epsilon,
+                                         layer._momentum, layer._weight_attr,
+                                         layer._bias_attr, layer._data_format,
+                                         layer._name)
+
+            if layer._weight_attr != False and layer._bias_attr != False:
```
Comment (XiaoguangHu01) [inline python/paddle/nn/layer/norm.py:1102]: False -> None

File: python/paddle/nn/layer/norm.py (hunk 7 above)
Comment (ceci3) [reply python/paddle/nn/layer/norm.py:570]: 如果weight_attr是False的话，self.weight就会是None，设置stop_gradient这一步就会报错..

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
